### PR TITLE
Person excluding should see exclusion of members

### PR DESCRIPTION
**Pull request by Powersource**
For https://github.com/ssbc/ssb-tribes/pull/72

### CODE REVIEW

**Comment by Powersource** [comment]
hmm might have to update keyring for this, if it doesn't already have a function for this. and if that's the case, i might have to first update keyring to the latest version. that should probably be a separate PR from this

**Comment by Powersource** [comment]
hmm from keyring v1 to v2 it seems to have lost the registerAuthors function. idk if it was mostly rewritten or what. wondering if i should reimplement that functionality in there or just calculate it in this module live instead. maybe that's the better option, starting with something unoptimized that works and doesn't change too much. should maybe still bump keyring

**Comment by Powersource** [comment]
yeah 1.1.0 only has registerAuthors, no unregister, so can't simply use that version https://gitlab.com/ahau/lib/ssb-keyring/-/tree/v1.1.0?ref_type=tags

File: method/group.js [hunk 1]
```diff
@@ -114,7 +114,15 @@ module.exports = function GroupMethods (ssb, keystore, state) {
 
       if (!excludeMemberSpec.isValid(content)) return cb(new Error(excludeMemberSpec.isValid.errorsString))
 
-      ssb.publish(content, cb)
+      ssb.publish(content, (err, msg)=> {
+        if (err) return cb(err)
+
+        // TODO: remove the excluded members from the keystore. the reverse of the 
+        //keystore.group.registerAuthors(groupId, authorIds, (err) => {
+        // that invite() does
```
Comment (mixmix) [inline method/group.js:122]: I think the only thing that might need to happen is consider undoing `ssb.replicate.request` ... but ... that should also depend on if they're a "friend" outside of this context

I think your work to not rely on keystore for queries is good.

File: test/api/exclude-members.test.js [hunk 2]
```diff
@@ -40,6 +40,10 @@ test('tribes.excludeMembers', async t => {
       recps: [groupId]
     }
     t.deepEqual(exclude.content, expected, 'kaitiaki excluded everyone')
+
+    const authors = await p(kaitiaki.tribes.listAuthors)(groupId)
+    console.log('authors', authors)
```
Comment (mixmix) [inline test/api/exclude-members.test.js:45]:
```suggestion
```

File: method/group.js [hunk 3]
```diff
@@ -126,18 +126,44 @@ module.exports = function GroupMethods (ssb, keystore, state) {
...
+            pull.filter(msg =>
+              groupId === msg.value.content.recps[0]
+            ),
+            pull.map(msg => msg.value.content.excludes),
+            pull.flatten(),
+            pull.unique(),
+            pull.collect((err, excludedMembers) => {
+              if (err) return cb(err)
+
+              const members = addedMembers.filter(addedMember => !excludedMembers.includes(addedMember))
```
Comment (Powersource) [inline method/group.js:162]: This disallows people from being added back into a group. That's probably possible to allow, but we might have to fix the member tangle first. Make issue for later (there's already an issue for the tangle, link to that)

File: method/group.js [hunk 4]
```diff
@@ -126,18 +126,44 @@ module.exports = function GroupMethods (ssb, keystore, state) {
...
+        ssb.query.read({ query: addedQuery }),
         pull.filter(addMemberSpec.isValid),
         pull.filter(msg =>
           groupId === msg.value.content.recps[0]
         ),
         pull.map(msg => msg.value.content.recps.slice(1)),
         pull.flatten(),
         pull.unique(),
-        pull.collect((err, members) => {
+        pull.collect((err, addedMembers) => {
```
Comment (Powersource) [inline method/group.js:137]: At first I was hoping to do some fancy pull stream stuff to get the streams to run in parallel, but after staring at the docs for a bit I decided that it's better to get to work simply to start with.

File: method/group.js (hunk 1 above)
Comment (Powersource) [reply method/group.js]: hmm good point. but i'm a bit afraid of ssb-friends. or is it that one doing the friend replication? doesn't look like ahau is using replication scheduler? in tribes2 i think both friends+group member replication is delegated to the scheduler, which makes it able to balance these two needs you bring up. should we ignore this problem for now maybe, make a new issue?

File: method/group.js (hunk 3 above)
Comment (mixmix) [reply method/group.js:162]: Agree with this approach. I think you should leave a comment in the code like 
```suggestion
              const members = addedMembers.filter(addedMember => !excludedMembers.includes(addedMember))
              // NOTE: this currently prevents people who've been removed from being re-added
```

**Comment by mixmix** [comment]
Made a squashed version on master here: https://github.com/ssbc/ssb-tribes/pull/84

**Pull request closed without merging**